src/cjss.js

```diff
@@ -1,93 +1,73 @@
-  const elements = document.querySelectorAll(selector);
+  const elements = selector === 'script' ? [null] : document.querySelectorAll(selector);
 
-  const js = getPureProperty(rule, '--js');
-  const html = getPureProperty(rule, '--html');
-  let data = getPureProperty(rule, '--data');
-
-  try {
-    data = JSON.parse(`{${ data }}`);
-  } catch (e) {
-    if (e instanceof SyntaxError) {
-      console.error(`CJSS: Invalid JSON found at ${ selector }: {${ data }}`);
-      console.error(e.message);
-      return false;
-    } else throw e;
-  }
-
-  if (html) {
-    for (const element of elements) {
-      const code = `return (\`${ html }\`)`;
-      const render = functionFromString(code, ['data', 'yield']);
+  const compiledStages = [];
 
+  for (const stage of Stage.ordered) {
+    const property = getProperty(rule, stage)
+      || (stage.fallback && getProperty(rule, stage.fallback));
+    if (property) {
+      const { mode, body } = property;
       try {
-        element.innerHTML = render.run([data, element.innerHTML]);
+        compiledStages.push({
+          body, mode, stage, run: getPlugin(stage, mode)(body),
+        });
       } catch (e) {
-        console.error('CJSS: Error in HTML:', e);
-        console.error(`at selector '${ selector }' and element`, element);
-        console.error(`of script:\n${ js }`);
-        return;
+        if (e instanceof CJSSError) {
+          console.error(e.message, e.original);
+          console.error(`in script ${stage}: ${mode}(${body})`);
+          // TODO: in the case of an error, should we skip this element or every element?
+          break;
+        } throw e;
       }
     }
   }
 
-  if (js) {
-    const jsRunner = functionFromString(js, ["data"]);
-
-    if (selector === 'script') {
+  for (const element of elements) {
+    let data = {};
+    for (const compiledStage of compiledStages) {
       try {
-        jsRunner.run([data]);
+        data = compiledStage.run(element, data) || data;
       } catch (e) {
-        console.error('CJSS: Error in JS:', e);
-        console.error(`at selector '${ selector }'`);
-        console.error(`of script:\n${ js }`);
-      }
-
-      return;
-    }
-
-    for (const element of elements) {
-      try {
-        jsRunner.run([data], element);
-      } catch (e) {
-        console.error('CJSS: Error in JS:', e);
-        console.error(`at selector '${ selector }' and element`, element);
-        console.error(`of script:\n${ js }`);
-        return;
+        if (e instanceof CJSSError) {
+          console.error(e.message, e.original);
+          if (element) console.error(`at selector '${selector}' and element`, element);
+          else console.error(`at selector '${selector}'`);
+          console.error(`in script ${compiledStage.stage}: ${
+            compiledStage.mode}(${compiledStage.body})`);
+          // TODO: in the case of an error, should we skip this element or every element?
```

Comment:
Same question here: after hitting an error when running one stage of the program, is logging the error, skipping this element and moving on to the next element good enough? In some cases this will result in the same error being logged repeatedly, as the same selector may apply to many elements, each of which suffer from the same error.

Reply:
You're welcome to try to find a way to refactor this nicely if you'd like. However, the only cases in which I am raising an exception are when exceptions have already been raised by the parsing or execution of user code, in which cases some exception handling is unavoidable.